### PR TITLE
iptables: ensure all rules are installed consistently

### PR DESCRIPTION
**Pull request by jibi**
See individual commits

### CODE REVIEW

**Comment by jibi** [comment]
/test-1.21-5.4

File: pkg/datapath/iptables/custom_chain.go [hunk 1]
```diff
@@ -94,87 +97,157 @@ var ciliumChains = []customChain{
...
+func (c *customChain) exists(prog iptablesInterface) (bool, error) {
+	args := []string{"-t", c.table, "-L", c.name}
+
+	output, err := prog.runProgCombinedOutput(args, true)
+	if err != nil {
+		if strings.Contains(string(output), "No chain/target/match by that name.") {
+			return false, nil
+		}
+
+		return false, fmt.Errorf("unable to list %s chain: %s (%s)", c.name, string(output), err)
```
Comment (christarazi) [inline pkg/datapath/iptables/custom_chain.go]: `%w` for errors

File: pkg/proxy/proxy.go [hunk 2]
```diff
@@ -289,9 +289,8 @@ func (p *Proxy) ackProxyPort(name string, pp *ProxyPort) error {
 		// Add rules for the new port
 		// This should always succeed if we have managed to start-up properly
 		scopedLog.Infof("Adding new proxy port rules for %s:%d", name, pp.proxyPort)
-		err := p.datapathUpdater.InstallProxyRules(pp.proxyPort, pp.ingress, name)
-		if err != nil {
-			return fmt.Errorf("Cannot install proxy rules for %s: %s", name, err)
+		if err := p.datapathUpdater.InstallProxyRules(pp.proxyPort, pp.ingress, name); err != nil {
+			return fmt.Errorf("Cannot install proxy rules for %s: %w", name, err)
```
Comment (christarazi) [inline pkg/proxy/proxy.go]: Nit: error strings should start lower-cased

File: pkg/datapath/iptables/iptables.go [hunk 3]
```diff
@@ -881,6 +883,28 @@ func (m *IptablesManager) InstallProxyRules(proxyPort uint16, ingress bool, name
...
+			log.WithError(err).Warning("Failed to install iptables proxy rules")
+			backoff.Wait(context.TODO())
+			continue
+		}
+
+		log.Info("Iptables proxy rules installed")
+		return nil
+	}
+
+	return nil
```
Comment (christarazi) [inline pkg/datapath/iptables/iptables.go]: Shouldn't this return the last error that occurred?

**Comment by jibi** [comment]
/test-1.23-net-next

**Comment by jibi** [comment]
/test-1.22-4.19

File: pkg/proxy/proxy.go [hunk 4]
```diff
@@ -289,9 +289,8 @@ func (p *Proxy) ackProxyPort(name string, pp *ProxyPort) error {
 		// Add rules for the new port
 		// This should always succeed if we have managed to start-up properly
 		scopedLog.Infof("Adding new proxy port rules for %s:%d", name, pp.proxyPort)
-		err := p.datapathUpdater.InstallProxyRules(pp.proxyPort, pp.ingress, name)
-		if err != nil {
-			return fmt.Errorf("Cannot install proxy rules for %s: %s", name, err)
+		if err := p.datapathUpdater.InstallProxyRules(pp.proxyPort, pp.ingress, name); err != nil {
```
Comment (christarazi) [inline pkg/proxy/proxy.go]: I think `Install{,Proxy}Rules` should take the daemon `context` so that we can stop the retries if they're in the middle of executing when the daemon is shutting down. 

See calls to `d.Datapath().Loader().Reinitialize()` for reference.

**Review by christarazi — CHANGES_REQUESTED** [review]
Quick first-glance review. Will take a deeper look when you mark it ready for review.

File: pkg/proxy/proxy.go (hunk 4 above)
Comment (jibi) [reply pkg/proxy/proxy.go]: added a `context` to both `InstallRules` and `InstallProxyRules` :+1: 
The latter was a bit involved as it's called also from `Proxy.ackProxyPort()`, which in turn is called from a few different places

File: pkg/datapath/iptables/custom_chain.go [hunk 5]
```diff
@@ -94,87 +97,157 @@ var ciliumChains = []customChain{
...
 	installMode := "-A"
 	if option.Config.PrependIptablesChains {
 		installMode = "-I"
 	}
 
+	feedRule := []string{"-m", "comment", "--comment", feederDescription + " " + c.name, "-j", c.name}
+	if feedArgs != "" {
+		argsList, err := shellwords.Parse(feedArgs)
+		if err != nil {
+			log.WithError(err).WithField(logfields.Object, feedArgs).Fatal("Unable to parse rule into argument slice")
```
Comment (christarazi) [inline pkg/datapath/iptables/custom_chain.go]: Fatal here generally seems out of place. Why does this failure get treated differently? AFAIK, we usually keep Fatal to the daemon subsystem.

File: pkg/datapath/iptables/iptables.go [hunk 6]
```diff
@@ -213,15 +233,8 @@ func (m *IptablesManager) removeCiliumRules(table string, prog iptablesInterface
...
-		for _, disabledChain := range option.Config.DisableIptablesFeederRules {
-			if strings.Contains(rule, " "+strings.ToUpper(disabledChain)+" ") {
-				log.WithField("chain", disabledChain).Info("Skipping the removal of feeder chain")
-				skipFeeder = true
-				break
-			}
-		}
-		if skipFeeder {
+		if skip, disabledChain := ruleReferencesDisabledChain(rule); skip {
+			log.WithField("chain", disabledChain).Info("Skipping the removal of feeder chain")
```
Comment (christarazi) [inline pkg/datapath/iptables/iptables.go]: Nit: let's add a logfield

**Review by christarazi — APPROVED** [review]

**Comment by jibi** [comment]
4.9 failure is legit:
```
➜  K8sDatapathConfig_IPv4Only_Check_connectivity_with_IPv6_disabled rg 'level=(fatal|err|warn)' cilium-combined-logs*
cilium-combined-logs.txt
874:2022-05-10T19:14:14.737756856Z level=warning msg="Failed to install iptables rules" error="unable to run 'ipset destroy cilium_node_set_v6' iptables command: ipset v7.5: Set cannot be destroyed: it is in use by a kernel component\n (exit status 1)" subsys=iptables
1049:2022-05-10T19:14:54.877748114Z level=warning msg="Failed to install iptables rules" error="unable to run 'ipset destroy cilium_node_set_v6' iptables command: ipset v7.5: Set cannot be destroyed: it is in use by a kernel component\n (exit status 1)" subsys=iptables
1922:2022-05-10T19:14:13.773727619Z level=warning msg="Failed to install iptables rules" error="unable to run 'ipset destroy cilium_node_set_v6' iptables command: ipset v7.5: Set cannot be destroyed: it is in use by a kernel component\n (exit status 1)" subsys=iptables
2149:2022-05-10T19:14:53.909722019Z level=warning msg="Failed to install iptables rules" error="unable to run 'ipset destroy cilium_node_set_v6' iptables command: ipset v7.5: Set cannot be destroyed: it is in use by a kernel component\n (exit status 1)" subsys=iptables
```

seems like there's still a rule referencing that ipset although ipv6 is disabled for the test (and so that rule should be removed):
```
➜  K8sDatapathConfig_IPv4Only_Check_connectivity_with_IPv6_disabled rg cilium_node_set_v6 bugtool-cilium-cvd58/cmd/ip6tables-save--c.md
15:[0:0] -A CILIUM_POST_nat -s fd02::100/120 -m set --match-set cilium_node_set_v6 dst -m comment --comment "exclude traffic to cluster nodes from masquerade" -j ACCEPT
```

looking into that

File: pkg/datapath/iptables/iptables.go [hunk 7]
```diff
@@ -438,148 +455,167 @@ func (m *IptablesManager) installStaticProxyRules() error {
...
 
-func (m *IptablesManager) doCopyProxyRules(prog iptablesInterface, table string, re *regexp.Regexp, match, oldChain, newChain string) {
-	output, err := prog.runProgCombinedOutput([]string{"-t", table, "-S"}, true)
+func (m *IptablesManager) doCopyProxyRules(prog iptablesInterface, table string, re *regexp.Regexp, match, oldChain, newChain string) error {
+	rules, err := prog.runProgCombinedOutput([]string{"-t", table, "-S"})
 	if err != nil {
-		log.WithFields(logrus.Fields{
-			"table": table,
-			"prog":  prog.getProg(),
-		}).WithError(err).Warning("Cannot list rules in table. Layer 7 proxy port may get reallocated, which could cause disruption for traffic selected by L7 policy", prog.getProg(), table)
```
Comment (pchaigno) [inline pkg/datapath/iptables/iptables.go:742]: Should we keep this information even if we don't emit a warning but return an error instead?

File: pkg/datapath/iptables/iptables.go [hunk 8]
```diff
@@ -590,40 +626,44 @@ func (m *IptablesManager) doCopyProxyRules(prog iptablesInterface, table string,
...
+
 			copyRule := append([]string{"-t", table}, args...)
-			log.WithField(logfields.Object, logfields.Repr(copyRule)).Debugf("Copying %s TPROXY rule from %s to %s", prog, oldChain, newChain)
-			err = prog.runProg(copyRule, true)
-			if err != nil {
-				log.WithFields(logrus.Fields{
-					"table":          table,
-					"prog":           prog.getProg(),
-					logfields.Object: rule,
-				}).WithError(err).Warn("Unable to copy TPROXY rule, disruption to traffic selected by L7 policy possible")
```
Comment (pchaigno) [inline pkg/datapath/iptables/iptables.go:767]: Same here. It seems we'll be loosing some context.

File: pkg/datapath/iptables/iptables.go [hunk 9]
```diff
@@ -590,40 +626,44 @@ func (m *IptablesManager) doCopyProxyRules(prog iptablesInterface, table string,
 				}).WithError(err).Warn("Unable to parse TPROXY rule, disruption to traffic selected by L7 policy possible")
 				continue
 			}
+
 			copyRule := append([]string{"-t", table}, args...)
-			log.WithField(logfields.Object, logfields.Repr(copyRule)).Debugf("Copying %s TPROXY rule from %s to %s", prog, oldChain, newChain)
```
Comment (pchaigno) [inline pkg/datapath/iptables/iptables.go:760]: Why was this removed?

File: pkg/datapath/iptables/iptables.go [hunk 10]
```diff
@@ -590,40 +626,44 @@ func (m *IptablesManager) doCopyProxyRules(prog iptablesInterface, table string,
...
+	return nil
 }
 
 // Redirect packets to the host proxy via TPROXY, as directed by the Cilium
 // datapath bpf programs via skb marks (egress) or DSCP (ingress).
 func (m *IptablesManager) addProxyRules(prog iptablesInterface, proxyPort uint16, ingress bool, name string) error {
-	output, err := prog.runProgCombinedOutput([]string{"-t", "mangle", "-S"}, true)
+	rules, err := prog.runProgCombinedOutput([]string{"-t", "mangle", "-S"})
 	if err != nil {
-		log.WithError(err).Warnf("Unable to list %s TPROXY rules: %s", prog, string(output))
```
Comment (pchaigno) [inline pkg/datapath/iptables/iptables.go:790]: We'll be loosing some context here as well, no?

File: pkg/datapath/iptables/iptables.go [hunk 11]
```diff
@@ -642,20 +682,20 @@ func (m *IptablesManager) addProxyRules(prog iptablesInterface, proxyPort uint16
...
+		if strings.Contains(rule, "-A CILIUM_PRE_mangle ") && !strings.Contains(rule, "cilium: TPROXY to host "+name) && strings.Contains(rule, portMatch) {
+
 			args, err := shellwords.Parse(strings.Replace(rule, "-A", "-D", 1))
 			if err != nil {
 				log.WithError(err).WithField(logfields.Object, rule).Warnf("Unable to parse %s TPROXY rule", prog)
 				continue
 			}
+
 			deleteRule := append([]string{"-t", "mangle"}, args...)
-			log.WithField(logfields.Object, logfields.Repr(deleteRule)).Debugf("Deleting stale %s TPROXY rule from mangle table", prog)
```
Comment (pchaigno) [inline pkg/datapath/iptables/iptables.go:821]: Why was this removed?

File: pkg/datapath/iptables/iptables_test.go [hunk 12]
```diff
@@ -107,6 +107,9 @@ func init() {
 func (s *iptablesTestSuite) TestRenameCustomChain(c *check.C) {
 	mockIp4tables := &mockIptables{c: c, prog: "iptables"}
 	mockIp4tables.expectations = []expectation{
+		{
+			args: "-t mangle -L CILIUM_PRE_mangle",
+		},
```
Comment (pchaigno) [inline pkg/datapath/iptables/iptables_test.go:112]: Why is this change needed? I expected this commit to have no functional impact on tests :thinking:

File: pkg/datapath/iptables/iptables.go [hunk 13]
```diff
@@ -850,6 +852,33 @@ func (m *IptablesManager) InstallProxyRules(proxyPort uint16, ingress bool, name
 	m.Lock()
 	defer m.Unlock()
 
+	backoff := backoff.Exponential{
+		Min:  20 * time.Second,
+		Max:  3 * time.Minute,
```
Comment (pchaigno) [inline pkg/datapath/iptables/iptables.go:863]: Doesn't this mean we'll hold the lock above for 3 minutes in the worst case? :grimacing:

File: pkg/datapath/iptables/iptables.go [hunk 14]
```diff
@@ -668,39 +668,37 @@ func (m *IptablesManager) addProxyRules(prog iptablesInterface, proxyPort uint16
...
+			continue
+		}
 
-			deleteRule := append([]string{"-t", "mangle"}, args...)
-			if err := prog.runProg(deleteRule); err != nil {
-				return err
-			}
+		deleteRule := append([]string{"-t", "mangle"}, args...)
+		if err := prog.runProg(deleteRule); err != nil {
+			return err
```
Comment (pchaigno) [inline pkg/datapath/iptables/iptables.go:717]: Seems like this change (from `iptables: refactor addProxyRules`) better belongs in the previous commit (`iptables: refactor removeCiliumRules`).

File: pkg/datapath/iptables/iptables.go [hunk 15]
```diff
@@ -614,23 +614,24 @@ func (m *IptablesManager) doCopyProxyRules(prog iptablesInterface, table string,
...
+				"table":          table,
+				"prog":           prog.getProg(),
+				logfields.Object: rule,
+			}).WithError(err).Warn("Unable to parse TPROXY rule, disruption to traffic selected by L7 policy possible")
+			continue
+		}
+
+		copyRule := append([]string{"-t", table}, args...)
+		if err := prog.runProg(copyRule); err != nil {
+			return err
```
Comment (pchaigno) [inline pkg/datapath/iptables/iptables.go:653]: This commit could probably be merged with the other that inverts conditions to simplify code. And instead of the usual `iptables: refactor X`, the title could then be `iptables: invert conditions to simplify logic`.

**Review by pchaigno — CHANGES_REQUESTED** [review]
A bunch of comments from me, but I don't think there's anything major apart maybe from the long lock holding.

I wonder if it would be best to split this into multiple PRs to ease backports if they are needed:
- First PR with the commits to ensure we error in case of... well errors :grin: 
- Second PR with the refactoring changes, optionally based on the first PR.
- Optionally, we could also move commits `iptables: always try to remove old IPv6 rules` and `iptables: remove old rules before deleting ipsets` to their own PRs if we think they may need to be backported as well.

File: pkg/datapath/iptables/iptables.go (hunk 9 above)
Comment (jibi) [reply pkg/datapath/iptables/iptables.go:760]: After adding the debug logs in `runProg()`/`runProgCombinedOutput()` I thought it would be redundant to keep these logs.

Sure, we'll see an error that only mentions the actual iptables command that failed rather than `Copying [..] TPROXY rule from $OLD_CHAIN to $CHAIN`, but that's probably not an issue as I think we can get already all the context we need just from the plain iptables command that failed (also, if we end up wrapping the error with the additional TPROXY context, why shouldn't we do that with every other rule?)

File: pkg/datapath/iptables/iptables.go (hunk 7 above)
Comment (jibi) [reply pkg/datapath/iptables/iptables.go:742]: I don't think the log is relevant anymore as now in case of failure we keep retrying instead of just ignoring the error (and we would also log an error message related to the failure of the list command)

File: pkg/datapath/iptables/iptables.go (hunk 10 above)
Comment (jibi) [reply pkg/datapath/iptables/iptables.go:790]: Same here: I think we should rely only on the debug logs in `runProg()`/`runProgCombinedOutput()`

File: pkg/datapath/iptables/iptables_test.go (hunk 12 above)
Comment (jibi) [reply pkg/datapath/iptables/iptables_test.go:112]: In order to not ignore any meaningful error, when we rename a chain we need to first test if the chain exists (so that we don't return an error in case it doesn't)

File: pkg/datapath/iptables/iptables.go (hunk 11 above)
Comment (jibi) [reply pkg/datapath/iptables/iptables.go:821]: (same reason)

File: pkg/datapath/iptables/iptables.go (hunk 13 above)
Comment (jibi) [reply pkg/datapath/iptables/iptables.go:863]: Correct, in the worst case no one else will be able to update the ruleset for the 3 minutes window (and the caller will block for that time). At the same time I'm not sure if making it async is a good strategy, as to my understanding the caller will assume that once this function returns the rules are in place (unless there are errors)

edit: synced on slack: will keep it sync for the first initialization and then make it async for subsequent calls to `InstallRules`/`InstallProxyRules` (but in a separate PR to avoid adding even more changes here)

File: pkg/datapath/iptables/custom_chain.go (hunk 5 above)
Comment (jibi) [reply pkg/datapath/iptables/custom_chain.go]: I do agree that fatal is a bit out of place here, but at the same time I was not sure what the best approach would be as we cannot recover from this error by simply rerunning `doInstallFeeder`, and if we just log it and continue, the iptables ruleset will be most certainly in an inconsistent state (so better fatal out loud and make the error condition clear).

What I can do is just return the error (to remain consistent with the rest of the logic here) and then let the daemon return a fatal error after the 3 attempts

File: pkg/datapath/iptables/iptables.go (hunk 14 above)
Comment (jibi) [reply pkg/datapath/iptables/iptables.go:717]: given that this is about refactoring `addProxyRules` code, I would leave it here

File: pkg/datapath/iptables/iptables.go (hunk 7 above)
Comment (pchaigno) [reply pkg/datapath/iptables/iptables.go:742]: Ok, but if we retry X times and eventually fail for good, what's the error message we will show? Will it highlight the potential disruption of traffic selected by L7 policies? (:point_left: that's the information I fear is now missing for users.)

Is there any downside to passing a new error with that message and the `err` instead of just `err`?

File: pkg/datapath/iptables/iptables.go (hunk 9 above)
Comment (pchaigno) [reply pkg/datapath/iptables/iptables.go:760]: > After adding the debug logs in `runProg()`/`runProgCombinedOutput()` I thought it would be redundant to keep these logs.

I agree. But when I was reviewing this commit, I didn't know about the last commit which actually has the related changes :wink: :wink: :wink:

File: pkg/datapath/iptables/iptables.go (hunk 14 above)
Comment (pchaigno) [reply pkg/datapath/iptables/iptables.go:717]: Fine by me, it's a nit really. I'll just note that organizing code changes according to which functions/files they change doesn't do anything to help understand/review.

File: pkg/datapath/iptables/iptables_test.go (hunk 12 above)
Comment (pchaigno) [reply pkg/datapath/iptables/iptables_test.go:112]: Oh, I see. Thanks!

File: pkg/datapath/iptables/iptables.go (hunk 13 above)
Comment (pchaigno) [reply pkg/datapath/iptables/iptables.go:863]: What if we moved the `m.Lock()` & `m.Unlock()` around `m.doInstallProxyRules()` inside the loop? Would there be any downside to that? It would avoid holding a lock while we're sleeping.

File: pkg/datapath/iptables/iptables.go (hunk 7 above)
Comment (jibi) [reply pkg/datapath/iptables/iptables.go:742]: > Is there any downside to passing a new error with that message and the err instead of just err?

no downside, I just added additional context about L7 policies traffic disruption to the error returned by `copyProxyRule`

File: pkg/datapath/iptables/iptables.go (hunk 13 above)
Comment (jibi) [reply pkg/datapath/iptables/iptables.go:863]: Good shout :+1: I can't see any downside of that as we are not accessing any `IptablesManager` state in `InstallRules`/`InstallProxyRules` so I moved the locks inside `doInstallRules`/`doInstallProxyRules`

**Review by pchaigno — APPROVED** [review]

File: pkg/datapath/iptables/iptables.go [hunk 16]
```diff
@@ -873,77 +805,85 @@ func (m *IptablesManager) endpointNoTrackRules(prog iptablesInterface, cmd strin
...
 }
 
-func (m *IptablesManager) InstallProxyRules(proxyPort uint16, ingress bool, name string) error {
+func (m *IptablesManager) InstallProxyRules(ctx context.Context, proxyPort uint16, ingress bool, name string) error {
+	backoff := backoff.Exponential{
+		Min:  20 * time.Second,
+		Max:  3 * time.Minute,
+		Name: "iptables-proxy-rules-installer",
+	}
+
```
Comment (ldelossa) [inline pkg/datapath/iptables/iptables.go:866]: nit: Any reason you don't make this a loop variable?

File: pkg/datapath/iptables/iptables.go [hunk 17]
```diff
@@ -404,6 +405,8 @@ var ciliumChains = []customChain{
 
 // IptablesManager manages the iptables-related configuration for Cilium.
 type IptablesManager struct {
+	lock.Mutex
+
```
Comment (jrajahalme) [inline pkg/datapath/iptables/iptables.go:268]: It would be good to document why this lock is needed while `IptablesManager` itself is not modified after `Init()` and why only some of the exported functions take the lock. Apparently this is due to maybe `Install` and `Remove` being called for the same underlying rules concurrently?

**Review by ldelossa — APPROVED** [review]
LGTM, a very dumb nit that you are free to ignore

**Review by jrajahalme — APPROVED** [review]
Generally it would ease backporting if bug fix commits would be separated out from refactoring to a PR of their own. With them interleaved like this it is easy to make a backporting error in one of the refactors if any of the moved code is different between versions.

It might be best to backport this PR separately from the normal backport process to avoid mistakes.

File: pkg/datapath/iptables/iptables.go (hunk 17 above)
Comment (jibi) [reply pkg/datapath/iptables/iptables.go:268]: I'll add a comment :+1: 

> Apparently this is due to maybe Install and Remove being called for the same underlying rules concurrently?

correct, we need this look to ensure there are no multiple instances of `InstallRules()`/`InstallProxyRules()` running concurrently as otherwise we may end up with errors or an inconsistent ruleset

**Comment by jibi** [comment]
Just updated a comment (CI was green beside the usual AKS failure), reviews are in, marking it as ready
